optaplanner-core/src/main/java/org/optaplanner/core/impl/score/inliner/ScoreInliner.java

```diff
@@ -16,18 +16,110 @@
+        ConstraintMatch<Score_> constraintMatch =
+                constraintMatchTotal.addConstraintMatch(justifications, score);
+        List<DefaultIndictment<Score_>> indictmentList = justifications.stream()
+                .distinct() // One match might have the same justification twice
+                .map(justification -> {
+                    DefaultIndictment<Score_> indictment = indictmentMap.computeIfAbsent(justification,
+                            __ -> new DefaultIndictment<>(justification, zeroScore));
+                    indictment.addConstraintMatch(constraintMatch);
+                    return indictment;
+                }).collect(Collectors.toList());
+        return () -> {
+            constraintMatchTotal.removeConstraintMatch(constraintMatch);
+            if (constraintMatchTotal.getConstraintMatchSet().isEmpty()) {
+                constraintMatchTotalMap.remove(constraintId);
+            }
+            for (DefaultIndictment<Score_> indictment : indictmentList) {
+                indictment.removeConstraintMatch(constraintMatch);
+                if (indictment.getConstraintMatchSet().isEmpty()) {
+                    indictmentMap.remove(indictment.getJustification());
+                }
+            }
+        };
+    }
+
+    public Map<String, ConstraintMatchTotal<Score_>> getConstraintMatchTotalMap() {
+        // Unchecked assignment necessary as CMT and DefaultCMT incompatible in the Map generics.
+        return (Map) constraintMatchTotalMap;
+    }
+
+    public Map<Object, Indictment<Score_>> getIndictmentMap() {
+        // Unchecked assignment necessary as Indictment and DefaultIndictment incompatible in the Map generics.
+        return (Map) indictmentMap;
+    }
+
+    protected void ensureNonZeroConstraintWeight(Score_ constraintWeight) {
+        if (constraintWeight.equals(zeroScore)) {
+            throw new IllegalArgumentException("The constraintWeight (" + constraintWeight + ") cannot be zero,"
+                    + " this constraint should have been culled during node creation.");
+        }
+    }
+
+    /**
+     * Runs constraint matching if enabled.
+     *
+     * Returns {@code undoWithoutConstraintMatch} if constraint matching is disabled.
+     * Otherwise adds a constraint match and extends the undo with an operation to remove the newly added match.
+     *
+     * @param constraintPackage never null
+     * @param constraintName never null
+     * @param constraintWeight never null
+     * @param undoWithoutConstraintMatch never null
+     * @param score never null, lazy creation ensures instance only created if constraint matching enabled
+     * @param justifications never null, lazy creation ensures collection only created if constraint matching enabled
+     * @return never null
+     */
+    protected UndoScoreImpacter buildUndo(String constraintPackage, String constraintName, Score_ constraintWeight,
```

Comment:
Let's make this method final to give the JVM a hint it can be inlined

Reply:
I'll remove it altogether.

This code is on the hot path, and even creating the supplier lambdas is overhead that doesn't need to be there when constraint matching is disabled. There will be a bit more code duplication, but the performance of this is really important.

Reply:
+1 I agree